### PR TITLE
docs: improve README

### PR DESCRIPTION
**Pull request by ielashi**
Added more information to the README:

* Background
* How it works
* Example Canister
* Links to other examples

The code snippets are all checked by cargo, so they're guaranteed to remain working.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,15 +1,141 @@
...
 
-  <p>
-    <a href="https://crates.io/crates/ic-stable-structures"><img alt="Crate Info" src="https://img.shields.io/crates/v/ic-stable-structures.svg"/></a>
-    <a href="https://github.com/dfinity/stable-structures/blob/master/LICENSE"><img alt="Apache-2.0" src="https://img.shields.io/github/license/dfinity/stable-structures"/></a>
-    <a href="https://docs.rs/ic-stable-structures"><img alt="API Docs" src="https://img.shields.io/badge/docs.rs-ic--stable--structures-blue"/></a>
-    <a href="https://forum.dfinity.org/"><img alt="Chat on the Forum" src="https://img.shields.io/badge/help-post%20on%20forum.dfinity.org-blue"></a>
-  </p>
-</div>
+Stable stuctures are designed to directly use stable memory as the backing store, allowing
+them to grow to GiBs in size without the need for `pre_upgrade`/`post_upgrade` hooks.
```
Comment (roman-kashitsyn) [inline README.md:13]:
```suggestion
Stable structures are designed to use stable memory as the backing store, allowing them to grow to gigabytes in size without the need for `pre_upgrade`/`post_upgrade` hooks.
```

File: README.md [hunk 2]
```diff
@@ -1,15 +1,141 @@
...
+
+## Background
+
+The conventional approach to canister state persistence is to serialize the entire state to stable memory in the pre_upgrade hook and decode it back in the post_upgrade hook. This approach is easy to implement and works well for relatively small datasets. Unfortunately, it does not scale well and can render a canister non-upgradable.
```
Comment (roman-kashitsyn) [inline README.md:17]: It's unclear how we split the text into lines in this document.
My favorite approach is to have one sentence per line; this dramatically simplifies looking at diffs and providing suggestions on code reviews.

```suggestion
The conventional approach to canister state persistence is to serialize the entire state to stable memory in the `pre_upgrade` hook and decode it back in the `post_upgrade` hook.
This approach is easy to implement and works well for relatively small datasets.
Unfortunately, it does not scale well and can render a canister non-upgradable.
```

File: README.md [hunk 3]
```diff
@@ -1,15 +1,141 @@
...
+Memories are abstracted with the [Memory] trait, and stable structures can work with any storage
+backend that implements this trait. This includes stable memory, a vector
+([VectorMemory]), or even a flat file ([FileMemory]).
+
+The example above initializes a [BTreeMap] with a [DefaultMemoryImpl], which maps to stable memory
+when used in a canister and to a [VectorMemory] otherwise.
+
+
+Note that **each stable structure needs a separate memory.** For example, this fails
+when run in a canister:
```
Comment (roman-kashitsyn) [inline README.md:54]:
```suggestion
Note that **stable structures cannot share memories.**
Each memory must belong to only one stable structure.
For example, this fails when run in a canister:
```

**Review by roman-kashitsyn — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (ielashi) [reply README.md]: Agreed. I made some more updates to make each sentence its own separate line. Thanks for the suggestion.

**Pull request merged**